mdata/cache/ccache_metric.go

```diff
@@ -0,0 +1,256 @@
+		if mc.endTs(keys[i]) > ts {
+			log.Debug("CCacheMetric seekAsc: seek found ts %d is between %d and %d", ts, keys[i], mc.endTs(keys[i]))
+			return keys[i], true
+		}
+	}
+
+	log.Debug("CCacheMetric seekAsc: seekAsc unsuccessful")
+	return 0, false
+}
+
+// assumes we already have at least a read lock
+func (mc *CCacheMetric) seekDesc(ts uint32, keys []uint32) (uint32, bool) {
+	log.Debug("CCacheMetric seekDesc: seeking for %d in the keys %+d", ts, keys)
+
+	for i := len(keys) - 1; i >= 0 && mc.endTs(keys[i]) > ts; i-- {
+		if keys[i] <= ts {
+			log.Debug("CCacheMetric seekDesc: seek found ts %d is between %d and %d", ts, keys[i], mc.endTs(keys[i]))
+			return keys[i], true
+		}
+	}
+
+	log.Debug("CCacheMetric seekDesc: seekDesc unsuccessful")
+	return 0, false
+}
+
+func (mc *CCacheMetric) searchForward(from, until uint32, keys []uint32, res *CCSearchResult) {
+	ts, ok := mc.seekAsc(from, keys)
+	if !ok {
+		return
+	}
+
+	// add all consecutive chunks to search results, starting at the one containing "from"
+	for ; ts != 0; ts = mc.chunks[ts].Next {
+		log.Debug("CCacheMetric searchForward: forward search adds chunk ts %d to start", ts)
+		res.Start = append(res.Start, mc.chunks[ts].Itgen)
+		endTs := mc.endTs(ts)
+		res.From = endTs
+
+		if endTs >= until {
+			res.Complete = true
+			break
+		}
+	}
+}
+
+func (mc *CCacheMetric) searchBackward(from, until uint32, keys []uint32, res *CCSearchResult) {
+	ts, ok := mc.seekDesc(until, keys)
+	if !ok {
+		return
+	}
+
+	for ; ts != 0; ts = mc.chunks[ts].Prev {
+		log.Debug("CCacheMetric searchBackward: backward search adds chunk ts %d to end", ts)
+		res.End = append(res.End, mc.chunks[ts].Itgen)
+		startTs := mc.chunks[ts].Ts
+		res.Until = startTs
+
+		if startTs <= from {
+			break
+		}
+	}
+}
+
+// the idea of this method is that we first look for the chunks where the
+// "from" and "until" ts are in. then we seek from the "from" towards "until"
+// and add as many cunks as possible to the result, if this did not result
+// in all chunks necessary to serve the request we do the same in the reverse
+// order from "until" to "from"
+// if the first seek in chronological direction already ends up with all the
+// chunks we need to serve the request, the second one can be skipped.
+
+// EXAMPLE:
+// from ts:                    |
+// until ts:                                                   |
+// cache:            |---|---|---|   |   |   |   |   |---|---|---|---|---|---|
+// chunks returned:          |---|                   |---|---|---|
+//
+func (mc *CCacheMetric) Search(res *CCSearchResult, from, until uint32) {
+	mc.RLock()
+	defer mc.RUnlock()
+
+	if len(mc.chunks) < 1 {
+		return
+	}
+
+	keys := mc.sortedTs()
+
+	mc.searchForward(from, until, keys, res)
+	if !res.Complete {
+		mc.searchBackward(from, until, keys, res)
+	}
+
+	if !res.Complete && res.From > res.Until {
+		fmt.Printf("found from > until (%d/%d), printing chunks\n", res.From, res.Until)
```

Comment:
what is the goal here? if it's to detect a situation that should not arise because it's incorrect and/or inferior because it will trigger cassandra lookups where they may not have been needed, then I think we need a good log prefix that will draw attention, e.g. `log.Warn()`